### PR TITLE
feat: added support for passing values to custom error messages with refine

### PR DESCRIPTION
**Pull request by teovillanueva**
Hey! First time contributing to the project. Feedback is appreciated :)

## Feature explanation

By allowing to pass values to the t function at the time of translating custom errors message we can create better translations. For example, at my company we have some custom rules that we've implemented like this.

```ts
// "errors.files.maxWeight": "Maximum file size is {{maxWeight}} Mb"

const fileWeight = z
    .number()
    .min(0)
    .refine((n) => n <= 5, {
      params: { i18n: { key: 'errors.files.maxWeight', values: { maxWeight: 5 }} },
    });

fileWeight.safeParse(12); // Maximum file size is 5 Mb
```

## Edit

You still can pass the translation key directly to the i18n param like before ;)
Also, some changes on this PR were made unintentionally by my formatter (prettier), it would probably be a good idea to add a `.prettierrc`

### CODE REVIEW

**Comment by igiagante** [comment]
Hi there. any progress of this? Is it possible to merge this PR?

File: packages/core/tests/makeZodI18nMap.test.ts [hunk 1]
```diff
@@ -234,5 +234,42 @@ describe("custom error message", () => {
...
+
+    z.setErrorMap(makeZodI18nMap({ ns: "zod_custom" }));
+
+    expect(
+      getErrorMessage(
+        z
+          .string()
+          .refine(() => false, {
+            params: {
+              i18n: { key: "test_custom_key", values: { myVal: "42069" } },
```
Comment (aiji42) [inline packages/core/tests/makeZodI18nMap.test.ts:268]: Please add tests for cases when `values` is null or not an object.

File: packages/core/src/index.ts [hunk 2]
```diff
@@ -222,7 +222,23 @@ export const makeZodI18nMap: MakeZodI18nMap = (option) => (issue, ctx) => {
...
+              key: i18nParam?.key ?? defaultKey,
+              values:
+                typeof i18nParam?.values === "object"
+                  ? (i18nParam?.values as object)
+                  : {},
+            }
+          : { key: "errors.custom", values: {} };
+
+      message = t(key, {
+        ...values,
```
Comment (aiji42) [inline packages/core/src/index.ts:241]: I would like to avoid casting with `as object`.  
Please define the following functions outside of `makeZodI18nMap`:

```ts
const isRecord = (value: unknown): value is Record<string, unknown> => {
  if (typeof value !== "object" || value === null) return false;

  for (const key in value) {
    if (!Object.prototype.hasOwnProperty.call(value, key)) return false;
  }

  return true;
};

const getKeyAndValues = (
  param: unknown,
  defaultKey: string
): {
  values: Record<string, unknown>;
  key: string;
} => {
  if (typeof param === "string") return { key: param, values: {} };

  if (isRecord(param)) {
    const key =
      "key" in param && typeof param.key === "string" ? param.key : defaultKey;
    const values =
      "values" in param && isRecord(param.values) ? param.values : {};
    return { key, values };
  }

  return { key: defaultKey, values: {} };
};
```

Then, you can simplify the code and strictly check `values` by doing this:

```ts
case ZodIssueCode.custom:
      const { key, values } = getKeyAndValues(
        issue.params?.i18n,
        "errors.custom"
      );
      message = t(key, {
        ...values,
        ns,
        defaultValue: message,
        ...path,
      });
      break;
```

**Review by aiji42 — CHANGES_REQUESTED** [review]
@teovillanueva Thank you for the cool feature addition. I have left comments, so please make the requested changes.

**Comment by teovillanueva** [comment]
Changes made! Let me know what you think :)

**Comment by aiji42** [comment]
Sorry for the late reply, dude. Looks good to me, gonna merge and release this.

@all-contributors please add @teovillanueva for code
